### DISCUSSION TITLE
Specify `event_hooks` at the `Request` level

### DISCUSSION THREAD

**Discussion by LBreedlove** [encode/httpx#2649] [General] [answered]
I'd like to be able to provide event_hooks in the Request object executed by the send methods. This would allow for providing per-request context data to the event_hooks, without instantiating a new client for each request. The request event_hooks would be executed in addition to the client level event_hooks.

I didn't see a way to do this in the current library; is there another feature providing this functionality, that I'm overlooking?

**Comment by lovelydinosaur** [comment, marked as answer]
Good question. Let's bump this into a discussion for now.
Can you start off by providing an example of what you'd like to be able to do, so we can work this through from a specific use-case?

**Reply by LBreedlove** [reply]
Basically, I'd like to be able to handle per-request notifications, specifically for capturing when the response has been received but not read from the network. I have a work around where I set `stream=True` when calling `send`, and then calling `response.read` after I've processed the response started event, but it would be nice to be able to use the builtin event_hooks instead.
If the request object has a context property (I haven't seen one?), that would be sufficient, since the response passed to the event hook has the request that initiated the response.

**Reply by lovelydinosaur** [reply]
So you can add some arbitrary per-request context using a request extension...

```python
def log_request(request):
    print(f"Request event hook: {request.method} {request.url} - Waiting for response")

def log_response(response):
    request = response.request
    context = request.extensions["some_extra_context_stuff"]
    print(f"Response event hook: {request.method} {request.url} - Status {response.status_code} - Context {context}")

client = httpx.Client(event_hooks={'request': [log_request])
response = client.get(..., extensions={"some_extra_context_stuff": ...})
```

Which might be enough for you to do whatever specifically per-request stuff it is within the event hook.

The extensions are currently only documented on `httpcore`... https://www.encode.io/httpcore/extensions/ - you wouldn't be using one of the pre-defined extensions, but instead using that namespace to include whatever context it is that you want to add.

**Reply by LBreedlove** [reply]
Ah, perfect, thank you! I wasn't aware of the extensions.

**Reply by lovelydinosaur** [reply]
Indeed. (https://github.com/encode/httpx/issues/2599)
